### PR TITLE
Update deoplete

### PR DESCRIPTION
**Pull request by pittcat**

### CODE REVIEW

**Comment by liuchengxu** [comment]
I'm fine with the `deoplete` related changes. But why `code-snippets` has to be updated in this PR?

**Comment by pittcat** [comment]
I think completion is related to code-snipets,and I want to make a common setting for all completion engine ,just like the code about tab maps

                function TabJumpNext()
                let snippet=UltiSnips#JumpForwards()
                if pumvisible()==0
                    if g:ulti_jump_forwards_res==1
                    return snippet
                    else
                    return "\<tab>"
                    endif
                endif
                endfunction
                inoremap <expr> <tab> pumvisible() ? "<C-n>" : "<C-R>=TabJumpNext()<CR>"

                function STabJumpBack()
                let snippet=UltiSnips#JumpBackwards()
                if pumvisible()==0
                    if g:ulti_jump_backwards_res==1
                    return snippet
                    else
                    return "\<s-tab>"
                    endif
                endif
                endfunction

                inoremap <expr> <s-tab> pumvisible() ? "<C-n>" : "<C-R>=TabJumpNext()<CR>"

                smap <TAB>   <Esc>:call UltiSnips#JumpForwards()<CR>
                smap <S-TAB> <Esc>:call UltiSnips#JumpBackwards()<CR>

But I think the enter key's map may be a little  different in different completion engine.I plan to place 
specific enter key's map to different completion engine layer.

deoplete layer

            "    enter key map
            let g:ulti_expand_res = 0
            function ExpandSnippetOrCarriageReturn()
                let snippet = UltiSnips#ExpandSnippet()
                if g:ulti_expand_res > 0
                    return snippet
                else 
                return deoplete#close_popup()
                endif
            endfunction

            inoremap <expr> <CR> pumvisible() ? "<C-R>=ExpandSnippetOrCarriageReturn()<CR>" : "\<CR>"

auto-complete layer

            
                " enter key setting
                function ExpandSnippetOrCarriageReturn()
                    let snippet = UltiSnips#ExpandSnippet()
                    if g:ulti_expand_res > 0
                        return snippet
                    else
                    return "\<C-y>"
                    endif
                endfunction

                inoremap <expr> <CR> pumvisible() ? "<C-R>=ExpandSnippetOrCarriageReturn()<CR>" : "\<CR>"

File: layers/+programming/code-snippets/config.vim [hunk 1]
```diff
@@ -6,20 +6,38 @@ scriptencoding utf-8
...
+     if g:ulti_jump_backwards_res==1
+       return snippet
+     else
+       return "\<s-tab>"
+     endif
+   endif
   endfunction
-  inoremap <expr> <CR> pumvisible() ? "\<C-R>=ExpandSnippetOrCarriageReturn()\<CR>" : "\<CR>"
+
+  inoremap <expr> <s-tab> pumvisible() ? "<C-n>" : "<C-R>=TabJumpNext()<CR>"
```
Comment (liuchengxu) [inline layers/+programming/code-snippets/config.vim]: typo `TabJumpNext()` -> `TabJumpBack()`.

File: layers/+completion/deoplete/config.vim [hunk 2]
```diff
@@ -1,139 +1,72 @@
...
+        let snippet = UltiSnips#ExpandSnippet()
+        if g:ulti_expand_res > 0
+            return snippet
+        else 
+          return deoplete#close_popup()
+        endif
+    endfunction
+
+    inoremap <expr> <CR> pumvisible() ? "<C-R>=ExpandSnippetOrCarriageReturn()<CR>" : "\<CR>"
+
```
Comment (liuchengxu) [inline layers/+completion/deoplete/config.vim]: Eliminate the extra blank line.

**Comment by liuchengxu** [comment]
When you eliminate the needless blank lines, then it looks good to me.

**Comment by liuchengxu** [comment]
Merged, thanks!

**Pull request merged**